include/toolchain.h

```diff
@@ -15,6 +15,24 @@
 #ifndef ZEPHYR_INCLUDE_TOOLCHAIN_H_
 #define ZEPHYR_INCLUDE_TOOLCHAIN_H_
 
+/**
+ * @def HAS_BUILTIN(x)
+ * @brief Check if the compiler supports the built-in function \a x.
+ *
+ * This macro is for use with conditional compilation to enable code using a
+ * builtin function that may not be available in every compiler.
+ */
+#ifdef __has_builtin
+#define HAS_BUILTIN(x) __has_builtin(x)
+#else
```

Comment:
Why bother here?  We're defining the HAS_BUILTIN_xxx symbols unconditionally anyway, right?  So what's the value of conditionally falling back to a toolchain macro when we could just do the same thing the same way for everyone and eliminate these lines?

Reply:
This case covers Clang-derived toolchains and [possibly future GCC versions](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970).

I'm trying to cover a more open-ended set of toolchains, a bit like Oticon's efforts.